jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,770 @@
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.pathmap.PathSpec;
+import org.eclipse.jetty.http.pathmap.PathSpecSet;
+import org.eclipse.jetty.server.ConnectionMetaData;
+import org.eclipse.jetty.server.Connector;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IncludeExclude;
+import org.eclipse.jetty.util.IncludeExcludeSet;
+import org.eclipse.jetty.util.InetAddressPattern;
+import org.eclipse.jetty.util.TypeUtil;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An abstract {@link Handler.Wrapper} that may conditionally handle a {@link Request}.
+ * The conditions are implemented by {@link IncludeExclude}s of:
+ * <ul>
+ *     <li>A method name, which can be efficiently matched</li>
+ *     <li>A {@link PathSpec} or sting representation, which can be efficient matched.</li>
+ *     <li>An arbitrary {@link Predicate} taking the {@link Request}, which is matched in a linear test of all predicates.</li>
+ * </ul>
+ * <p>If the conditions are met, the abstract {@link #doHandle(Request, Response, Callback)} method will be invoked,
+ * otherwise the {@link #doNotHandle(Request, Response, Callback)} method will be invoked, which by default calls
+ * the {@link #nextHandle(Request, Response, Callback)} method.</p>
+ * <p>A typical usage is to extend the class and provide an implementation of {@link #doHandle(Request, Response, Callback)}:</p>
+ * <pre>{@code
+ * public class MyOptionalHandler extends ConditionalHandler
+ * {
+ *     @Override
+ *     public boolean doHandle(Request request, Response response, Callback callback)
+ *     {
+ *         response.getHeaders().add("Test", "My Optional Handling");
+ *         return nextHandle(request, response, callback);
+ *     }
+ * }
+ * }</pre>
+ * <p>If the conditions added to {@code MyOptionalHandler} are met, then the {@code doHandle} method is called
+ * and a response header added before invoking {@link #nextHandle(Request, Response, Callback)}, otherwise
+ * the {@link #nextHandle(Request, Response, Callback)} is directly invoked.</p>
+ *
+ * <p>Alternately, one of the concrete subclasses may be used with the optional behaviour in the following {@link Handler}(s).
+ * These implementations all call {@link #nextHandle(Request, Response, Callback)} if the conditions are met, otherwise
+ * they vary in how they behave in {@link #doNotHandle(Request, Response, Callback)} when the conditions are not met:</p>
+ * <ul>
```

Comment:
I think that the logic would be more natural  iff these concrete handlers meet the conditions, then their doHandle() method is called - just like the `ConditionalHandler` - and  does the special behaviour (skipping, forbidding etc). Their `doNotHandle()` method would only be called iff the conditions are not met, just falling throught to the `ConditionalHandler`.doNotHandle() method to call `nextHandle()`.

Reply:
Ooooooohhhh!   I think that is what @sbordet was saying when he said: "A class should always be named after what it does"!!!

Yeah that makes more sense.... we Forbid if conditions are met, not the otherway around!

Got it.  Will do

Reply:
done

Reply:
I did not mean that, and I think it's confusing to have these classes do their action when the conditions are met, they do the action when the conditions are not met.

Because it's more likely to think "do when the method is ... and the path is ..." so the natural thinking is for includes, not excludes.
It's when the condition is not met that I want to reject the request (or not handle it, or skip).

Reply:
@sbordet I think it is confusing for these handlers to not do their actions when their conditions are met.